### PR TITLE
Update existing solvers to new interface

### LINKED ISSUE #956: Move existing solvers to new interface
## Description of issue / requirement to address

Following #928, the existing solvers should be updated to comply to the new `SolverABC` interface.

The solvers to update are:

- `FileProgramInterface`
- `plasmod.solver`
- `process.solver`

### PR DESCRIPTION
**Pull request by hsaunders1904**
## Linked Issues

Closes #956 

## Description

Migrates the Plasmod and PROCESS solvers to the new solver interface. This is a feature branch being merged, so is a combination of several other PRs:

- #1054 
- #1146 
- #1167 
- #1169 

## Interface Changes

PROCESS and Plasmod solvers.

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `flake8` and `black .`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/codes/process/_solver.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        self.read_directory = _build_config.pop("read_dir", os.getcwd())
+        self.template_in_dat = _build_config.pop("template_in_dat", DEFAULT_INDAT)
+        self.problem_settings = _build_config.pop("problem_settings", {})
+        self.in_dat_path = _build_config.pop(
+            "in_dat_path", os.path.join(self.run_directory, "IN.DAT")
+        )
+        if len(_build_config) > 0:
+            bluemira_warn(
+                f"'{self.name}' solver received unknown build_config arguments: "
+                f"'{''', '''.join(_build_config.keys())}'."
```
Comment (je-cook) [inline bluemira/codes/process/_solver.py:141]: Extra quotes seemingly for no reason? This produces the same result for me:
```suggestion
                f"'{', '.join(_build_config.keys())}'."
```

**Review by je-cook — APPROVED** [review]
Just one very minor but all good from my end

File: bluemira/codes/process/_solver.py (hunk 1 above)
Comment (je-cook) [reply bluemira/codes/process/_solver.py:141]: How about this:

```python
In [6]: d = {"a":1, "b":2, "c":3}

In [7]: block = "', '"

In [8]: f"'{block.join(d.keys())}'."
Out[8]: "'a', 'b', 'c'."

```

Comment (hsaunders1904) [reply bluemira/codes/process/_solver.py]: Done 👍. 

Another cool way of doing this I came across: `str(list(d.keys()))[1:-1]`. Less easy to read than your solution though,

**Pull request merged**